### PR TITLE
Add new Adoppler Bidder

### PR DESCRIPTION
**Pull request by AndriyPavlyuk**
Add new Adoppler Bidder to keep Up with Go version.

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                        .singletonList(imp)).build();
+                final String body = mapper.encode(outgoingRequest);
+                final MultiMap headers = HttpUtil.headers().add("x-openrtb-version", "2.5");
+                result.add(HttpRequest.<BidRequest>builder()
+                        .method(HttpMethod.POST)
+                        .uri(endpointUrl)
+                        .headers(headers)
+                        .payload(outgoingRequest)
+                        .body(body)
+                        .build());
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: please extract this in a separate method `createSingleRequest`

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                final ExtImpAdoppler validExtImp = parseAndValidateImpExt(imp);
+                final String updateRequestId = request.getId() + "-" + validExtImp.getAdunit();
+                final BidRequest updateRequest = request.toBuilder().id(updateRequestId).build();
+                final BidRequest outgoingRequest = updateRequest.toBuilder().imp(Collections
+                        .singletonList(imp)).build();
+                final String body = mapper.encode(outgoingRequest);
+                final MultiMap headers = HttpUtil.headers().add("x-openrtb-version", "2.5");
+                result.add(HttpRequest.<BidRequest>builder()
+                        .method(HttpMethod.POST)
+                        .uri(endpointUrl)
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: Where the logic related to custom url in which they use `ext.adUnit`

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            if (imp.getBanner() != null) {
+                impTypes.put(imp.getId(), BidType.banner);
+            } else if (imp.getVideo() != null) {
+                impTypes.put(imp.getId(), BidType.video);
+            } else if (imp.getAudio() != null) {
+                impTypes.put(imp.getId(), BidType.audio);
+            } else if (imp.getXNative() != null) {
+                impTypes.put(imp.getId(), BidType.xNative);
+            } else {
+                return Result.emptyWithError(BidderError.badInput(String.format(
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: redundant `.format`

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                        "one of $.imp.banner, $.imp.video, $.imp.audio and $.imp.native field required")));
+            }
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: I'd probably extracted this into separate method

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                final AdopplerResponseExt adopplerResponseExt;
+                if (impTypes.get(bid.getImpid()) == BidType.video) {
+                    final ObjectNode ext = bid.getExt();
+                    try {
+                        adopplerResponseExt = parseResponseExt(ext);
+                    } catch (PreBidException e) {
+                        return Result.emptyWithError(BidderError.badServerResponse(e.getMessage()));
+                    }
+                    final AdopplerResponseVideoExt adopplerResponseVideoExt = adopplerResponseExt.getAds();
+                    if ((adopplerResponseExt == null) || (adopplerResponseVideoExt == null)) {
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: (adopplerResponseExt == null) redundant (or in the wrong place)

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                        return Result.emptyWithError(BidderError.badServerResponse(String
+                                .format("$.seatbid.bid.ext.ads.video required")));
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: .format redundant

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                    }
+                    ExtBidPrebidVideo.of(adopplerResponseVideoExt.getVideo().getDuration(), head(bid.getCat()));
+                }
+                final BidderBid bidderBid = BidderBid.of(bid, impTypes.get(bid.getImpid()), DEFAULT_BID_CURRENCY);
+                bidderBids.add(bidderBid);
+            }
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: I'd extract this into 2 separate methods 
1 one return List<BidderBid>
2 get Type and return ExtBidPrebidVideo

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+                        adopplerResponseExt = parseResponseExt(ext);
+                    } catch (PreBidException e) {
+                        return Result.emptyWithError(BidderError.badServerResponse(e.getMessage()));
+                    }
+                    final AdopplerResponseVideoExt adopplerResponseVideoExt = adopplerResponseExt.getAds();
+                    if ((adopplerResponseExt == null) || (adopplerResponseVideoExt == null)) {
+                        return Result.emptyWithError(BidderError.badServerResponse(String
+                                .format("$.seatbid.bid.ext.ads.video required")));
+                    }
+                    ExtBidPrebidVideo.of(adopplerResponseVideoExt.getVideo().getDuration(), head(bid.getCat()));
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: As you can see, we do not have separate place for BidVideo. (thats why this code is unnecessary) But you can exclude it into separate methods to use it in a future

Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: .getVideo can produce NPE.

File: src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        assertThat(result.getErrors().get(0).getMessage())
+                .startsWith("one of $.imp.banner, $.imp.video, $.imp.audio and $.imp.native field required");
+        assertThat(result.getErrors().get(0).getType()).isEqualTo(BidderError.Type.bad_input);
+        assertThat(result.getValue()).isEmpty();
+    }
+
+    @Test
+    public void makeBidsShouldReturnErrorIfBidIdEmpty() throws JsonProcessingException {
+        // given
+        BidRequest bidRequest = BidRequest.builder()
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java]: final

File: src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+        assertThat(result.getErrors().get(0).getMessage())
+                .startsWith("duplicate $.imp.id impId");
+        assertThat(result.getErrors().get(0).getType()).isEqualTo(BidderError.Type.bad_input);
+        assertThat(result.getValue()).isEmpty();
+    }
+
+    @Test
+    public void makeBidsShouldReturnErrorIfEmptyImp() throws JsonProcessingException {
+        // given
+        BidRequest bidRequest = BidRequest.builder()
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java]: final

File: src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    @Test
+    public void makeBidsShouldReturnErrorIfDuplicateId() throws JsonProcessingException {
+        // given
+        Imp imp1 = Imp.builder().id("impId").banner(Banner.builder().build()).build();
+        Imp imp2 = Imp.builder().id("impId").video(Video.builder().build()).build();
+        List imps = new ArrayList();
+        imps.add(imp1);
+        imps.add(imp2);
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java]: final

File: src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+        assertThat(result.getValue()).isEmpty();
+    }
+
+    @Test
+    public void makeBidsShouldReturnErrorIfExtEmpty() throws JsonProcessingException {
+        // given
+        Imp imp = Imp.builder().id("impId").video(Video.builder().build()).build();
+        List imps = new ArrayList();
+        imps.add(imp);
+        BidRequest bidRequest = BidRequest.builder().imp(imps).build();
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java]: final

File: src/test/java/org/prebid/server/it/AdopplerTest.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+                .header("Origin", "http://www.example.com")
+                .cookie("uids", "eyJ1aWRzIjp7ImdhbW9zaGkiOiJHTS1VSUQifX0=")
+                .body(jsonFrom("openrtb2/adoppler/test-auction-adoppler-request.json"))
+                .post("/openrtb2/auction");
+
+        // then
+        final String expectedAuctionResponse = openrtbAuctionResponseFrom(
+                "openrtb2/adoppler/test-auction-adoppler-response.json",
+                response, singletonList("adoppler"));
+        String actualStr = response.asString();
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/it/AdopplerTest.java]: final

File: src/test/resources/org/prebid/server/it/openrtb2/adoppler/test-auction-adoppler-request.json [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        "minduration": 120,
+        "mimes": [
+          "video/mp4"
+        ]
+      },
+      "ext": {
+        "adoppler": {
+          "adunit": "unit2"
+        }
+      }
```
Comment (DGarbar) [inline src/test/resources/org/prebid/server/it/openrtb2/adoppler/test-auction-adoppler-request.json]: duplicate

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java (hunk 4 above)
Comment (AndriyPavlyuk) [reply src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java]: I think It is not good idea due to return statements

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 15]
```diff
@@ -60,17 +61,8 @@ public AdopplerBidder(String endpointUrl, JacksonMapper mapper) {
...
-                final String body = mapper.encode(outgoingRequest);
-                final MultiMap headers = HttpUtil.headers().add("x-openrtb-version", "2.5");
-                result.add(HttpRequest.<BidRequest>builder()
-                        .method(HttpMethod.POST)
-                        .uri(endpointUrl)
-                        .headers(headers)
-                        .payload(outgoingRequest)
-                        .body(body)
-                        .build());
+                final String url = endpointUrl + "/processHeaderBid/" + validExtImp.getAdunit();
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java:64]: Notice that they used `url.PathEscape` so you need to use appropriate method in HttpUtils for .getAnunit

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 16]
```diff
@@ -110,11 +115,38 @@ private ExtImpAdoppler parseAndValidateImpExt(Imp imp) {
             return Result.emptyWithError(BidderError.badServerResponse(String.format("invalid body: %s",
                     e.getMessage())));
         }
+        final Map<String, BidType> impTypes;
+        final List<BidderBid> bidderBids;
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java:119]: redundant

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 17]
```diff
@@ -110,11 +115,38 @@ private ExtImpAdoppler parseAndValidateImpExt(Imp imp) {
...
+        try {
+            impTypes = getImpTypes(bidRequest);
+            bidderBids = getBids(bidResponse, impTypes);
+        } catch (PreBidException e) {
+            return Result.emptyWithError(BidderError.badInput(e.getMessage()));
+        }
+        return Result.of(bidderBids, Collections.emptyList());
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java:126]: can be in try block

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 18]
```diff
@@ -110,11 +115,38 @@ private ExtImpAdoppler parseAndValidateImpExt(Imp imp) {
             return Result.emptyWithError(BidderError.badServerResponse(String.format("invalid body: %s",
                     e.getMessage())));
         }
+        final Map<String, BidType> impTypes;
+        final List<BidderBid> bidderBids;
+        try {
+            impTypes = getImpTypes(bidRequest);
+            bidderBids = getBids(bidResponse, impTypes);
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java:122]:
```
 try {
            bidResponse = decodeBodyToBidResponse(httpCall);
        } catch (PreBidException e) {
            return Result.emptyWithError(BidderError.badServerResponse(String.format("invalid body: %s",
                    e.getMessage())));
        }
```
Can use same try block 
And code will be much cleaner

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 19]
```diff
@@ -110,11 +115,38 @@ private ExtImpAdoppler parseAndValidateImpExt(Imp imp) {
...
+        }
+    }
+
+    private AdopplerResponseExt parseResponseExt(ObjectNode ext) {
+        try {
+            return mapper.mapper().treeToValue(ext, AdopplerResponseExt.class);
+        } catch (JsonProcessingException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+    }
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java:143]: This is used later in the code

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 20]
```diff
@@ -125,54 +157,46 @@ private ExtImpAdoppler parseAndValidateImpExt(Imp imp) {
...
-                    }
-                    final AdopplerResponseVideoExt adopplerResponseVideoExt = adopplerResponseExt.getAds();
-                    if ((adopplerResponseExt == null) || (adopplerResponseVideoExt == null)) {
-                        return Result.emptyWithError(BidderError.badServerResponse(String
-                                .format("$.seatbid.bid.ext.ads.video required")));
-                    }
-                    ExtBidPrebidVideo.of(adopplerResponseVideoExt.getVideo().getDuration(), head(bid.getCat()));
+                    throw new PreBidException(String.format("unknown impid: %s", bid.getImpid()));
                 }
+                setExtBidPrebidVideo(bid, impTypes);
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java:174]: You did not set anything. You just return this value. Rename method.

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java (hunk 18 above)
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java:122]: getBids Can be replaced
```
 bidResponse.getSeatbid().stream()
                .filter(Objects::nonNull)
                .map(SeatBid::getBid)
                .filter(Objects::nonNull)
                .flatMap(Collection::stream)
                .map(bid -> TODO createBid())
                .collect(Collectors.toList())
```

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 21]
```diff
@@ -125,54 +157,46 @@ private ExtImpAdoppler parseAndValidateImpExt(Imp imp) {
...
-            throw new PreBidException(e.getMessage(), e);
+    private ExtBidPrebidVideo setExtBidPrebidVideo(Bid bid, Map<String, BidType> impTypes) {
+        final AdopplerResponseExt adopplerResponseExt;
+        if (impTypes.get(bid.getImpid()) == BidType.video) {
+            final ObjectNode ext = bid.getExt();
+            try {
+                adopplerResponseExt = parseResponseExt(ext);
+            } catch (PreBidException e) {
+                throw new PreBidException(e.getMessage());
+            }
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java:190]: Do you really need catch to throw SIMILAR exception ?

File: src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java [hunk 22]
```diff
@@ -84,7 +84,7 @@ public void makeHttpRequestsShouldCreateCorrectURL() {
         // then
         assertThat(result.getErrors()).isEmpty();
         assertThat(result.getValue()).hasSize(1);
-        assertThat(result.getValue().get(0).getUri()).isEqualTo(ENDPOINT_URL);
+        assertThat(result.getValue().get(0).getUri()).isEqualTo(ENDPOINT_URL+"/processHeaderBid/adUnit");
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java:87]: style
+ cant you just add `/processHeaderBid/adUnit` to ENDPOINT_URL ?

File: src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java [hunk 23]
```diff
@@ -190,17 +190,17 @@ public void makeBidsShouldReturnErrorIfBidIdEmpty() throws JsonProcessingExcepti
...
         assertThat(result.getValue()).isEmpty();
     }
 
     @Test
     public void makeBidsShouldReturnErrorIfExtEmpty() throws JsonProcessingException {
         // given
-        Imp imp = Imp.builder().id("impId").video(Video.builder().build()).build();
-        List imps = new ArrayList();
+        final Imp imp = Imp.builder().id("impId").video(Video.builder().build()).build();
+        final List imps = new ArrayList();
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java:201]: use singletonList

File: src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java (hunk 22 above)
Comment (AndriyPavlyuk) [reply src/test/java/org/prebid/server/bidder/adoppler/AdopplerBidderTest.java:87]: That will be incorrect

File: src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+
+    private BidderBid createBid(Bid bid, Map<String, BidType> impTypes) {
+        if (impTypes.get(bid.getImpid()) == null) {
+            throw new PreBidException(String.format("unknown impid: %s", bid.getImpid()));
+        }
+        getExtBidPrebidVideo(bid, impTypes);
+        return BidderBid.of(bid, impTypes.get(bid.getImpid()), DEFAULT_BID_CURRENCY);
+    }
+
+    private ExtBidPrebidVideo getExtBidPrebidVideo(Bid bid, Map<String, BidType> impTypes) {
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java:173]: What is the reason for this method?
The returned value is not used anywhere.

**Pull request merged**